cmake/compilerFlags.cmake

```diff
@@ -20,17 +20,28 @@ if ( MINGW OR UNIX OR MSYS ) # MINGW, Linux, APPLE, CYGWIN
         endif()
     endif()
 
+    if (COMPILER_IS_GCC AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 8.0)
+        add_compile_options(-fstack-clash-protection -fcf-protection)
+    endif()
+
     if (COMPILER_IS_GCC OR COMPILER_IS_CLANG)
+        add_compile_options(-Wp,-D_GLIBCXX_ASSERTIONS)
+
+        if (CMAKE_BUILD_TYPE STREQUAL Release AND NOT APPLE)
+            add_compile_options(-Wp,-D_FORTIFY_SOURCE=2) # Requires to compile with -O2
+        endif()
 
         if(BUILD_WITH_COVERAGE)
-            set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g ")
-            set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O0")
-            set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fprofile-arcs")
-            set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -ftest-coverage")
+            add_compile_options(--coverage)
+            # TODO: From CMake 3.13 we could use add_link_options instead these 2 lines
             set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} --coverage")
+            set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} --coverage")
         endif()
 
-        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wall -Wcast-align -Wpointer-arith -Wformat-security -Wmissing-format-attribute -Woverloaded-virtual -W")
+        add_compile_options(-Wall -Wcast-align -Wpointer-arith -Wformat-security -Wmissing-format-attribute -Woverloaded-virtual -W)
+        add_compile_options(-fstack-protector-strong)

```

Comment:
Note that this fails the gitlab CI on CentOS as `-fstack-protector-strong` is not available for clang 3.4.2. Unfortunately I don't know when it got added, but it appears to be present in clang 3.7.

Reply:
Wow ... it actually also failed in Fedora_MinGW:

```
[  2%] Building CXX object xmpsdk/CMakeFiles/exiv2-xmp.dir/src/UnicodeConversions.cpp.obj
during RTL pass: final
/builds/D4N/exiv2/xmpsdk/src/UnicodeConversions.cpp: In function 'void ToUTF16(const UTF8Unit*, size_t, std::__cxx11::string*, bool)':
/builds/D4N/exiv2/xmpsdk/src/UnicodeConversions.cpp:293:1: internal compiler error: in i386_pe_seh_unwind_emit, at config/i386/winnt.c:1258
 } // ToUTF16
```

No idea what this is about ... 

I will add the additional guard for CentOS

Reply:
Yeah, I saw that. I **really** hope that that error is just transient and will maybe disappear…

Could be that these hardening flags don't mix with MinGW well, because cross compilation is weird and windows is too…

@cgilles have you ever seen the above error?